src/net.cpp

```diff
@@ -1858,13 +1858,14 @@ void CConnman::ThreadMnbRequestConnections()
         std::pair<CService, std::set<uint256> > p = mnodeman.PopScheduledMnbRequestConnection();
         if(p.first == CService() || p.second.empty()) continue;
 
-        CNode* pnode = NULL;
-        {
-            LOCK2(cs_main, cs_vNodes);
-            pnode = ConnectNode(CAddress(p.first, NODE_NETWORK), NULL, true);
-            if(!pnode) continue;
-            pnode->AddRef();
-        }
+        CNode* pnode = ConnectNode(CAddress(p.first, NODE_NETWORK), NULL, true);
+
+        // To ensure that node wasn't/won't be moved to "disconnected"
+        // and/or wasn't/won't be deleted in ThreadSocketHandler
+        LOCK(cs_vNodes);
+
+        // Too late, node was either deleted or disconnected before we locked above
+        if(!pnode || pnode->fDisconnect) continue;
```

Comment:
At this point, pnode would be a stale pointer if it got deleted already. So the "!pnode" check would still be true, but fDisconnect would contain stale data.

What about ignoring the return value of ConnectNode here and do a FindNode with cs_vNodes locked afterwards. This way the lock is only held in a very short timespan and when checking the return value of FindNode, you can be sure it did not get deleted. An alternative would be to use ForNode instead, but I'm not sure about where the grant.MoveTo had to be put in that case, as I didn't check what it is for.